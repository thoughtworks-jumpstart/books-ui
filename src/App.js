import React, { Component } from "react";
import "./App.css";

const API_HOST = process.env.REACT_APP_BOOKS_API || "http://localhost:3000";

class DataFetcher extends Component {
  state = {
    data: [],
    isLoading: true
  };

  async componentDidMount() {
    const url = `${API_HOST}${this.props.endpoint}`;
    try {
      const response = await fetch(url);
      if (response.ok) {
        const data = await response.json();
        this.setState({
          data
        });
      }
    } catch (err) {
      console.err(err);
    } finally {
      this.setState({
        isLoading: false
      });
    }
  }

  render() {
    if (this.state.isLoading) {
      return <div>Loadingâ€¦</div>;
    }
    return this.props.render(this.state);
  }
}

const Books = ({ data = [] }) => {
  return (
    <div>
      <h1>Books</h1>
      {data.map(book => {
        return <li key={book._id}>{book.title}</li>;
      })}
    </div>
  );
};

const Authors = ({ data = [] }) => {
  return (
    <div>
      <h1>Authors</h1>
      {data.map(author => {
        return <li key={author._id}>{author.name}</li>;
      })}
    </div>
  );
};
class App extends Component {
  render() {
    return (
      <div>
        <DataFetcher endpoint="/books" render={props => <Books {...props} />} />
        <DataFetcher
          endpoint="/authors"
          render={props => <Authors {...props} />}
        />
      </div>
    );
  }
}

export default App;
