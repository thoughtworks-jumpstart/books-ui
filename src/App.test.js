import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import App from "./App";

it("renders without crashing", () => {
  const renderer = new ShallowRenderer();
  renderer.render(<App />);
  const output = renderer.getRenderOutput();
  expect(output).toMatchSnapshot();
});
