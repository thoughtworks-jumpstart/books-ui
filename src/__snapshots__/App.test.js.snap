// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <DataFetcher
    endpoint="/books"
    render={[Function]}
  />
  <DataFetcher
    endpoint="/authors"
    render={[Function]}
  />
</div>
`;
